README.md
# boa
boa web server


src/Changelog
** Changes from 0.94.8 to 0.94.8.1
 * Change umask call from (umask(0600)) to (umask(~0600))

** Changes from 0.94.7 to 0.94.8 
 * Fix major thinko in temp file permissions
 * unlink temporary file immediately following creation
 * implement maximum # of active connections at 10 less than RLIMIT_NOFILE
   to avoid or eliminate crashes resulting from running out of 
   file descriptors
 * Fix thinko in POST
 
** Changes from 0.94.6 to 0.94.7
 * STDIN and STDOUT are now tied to /dev/null
 * sets PATH_MAX to 2048 if not defined (for Hurd)
 * core dumps (should never happen) would be located in /tmp
 * alter behavior when select gets a EBADF
 * add translation for the \" char -> &quot;
 * remove use of sys_errlist.  Use perror.
 * better makedist.sh (still a stupid program though)

** Changes from 0.94.5 to 0.94.6
 * Removed doc++ commenting
 * Removed erroneous debugging statments
 * Move some stuff out of config.c (read_config_file) to boa.c
 * Altered some of fixup_server_root()
 * Bug fix in get.c re: automatic gunzip
 * Added some stubs for chroot code (*not* ready yet)

** Changes from 0.94.4 to 0.94.5
 * Alteration of most of the comments and such for doc++ use
 * Fixed buffer overflow in alias.c
 * Fixed buffer underflow in util.c

** Changes from 0.94.3 to 0.94.4
 * Better escaping of data to user, both for HTTP headers and HTML body
 * Proper escaping of output in CGI example perl scripts

** Changes from 0.94.0 to 0.94.2
 * Fixed obnoxious pipeline bug
 * Fixed (sorta) a compilation/core bug for *BSD systems 
   Original code by Thomas Neumann <tn@tmr-online.de>
 * Moved to GPLv2
 * Changed manpage to section 8
 * boa.sgml now references a .png file instead of evil .gif

** Changes from 0.93.19.2 to 0.94.0
 * Added UseGMT to the configuration parser
 * util.c commonlog now logs in Apache-style commonlog time format
 * Remove SO_SNDBUF on-start message

** Changes from 0.93.19 to 0.93.19.2
 * Changed to combined log (from NCSA access_log format) ala Drew Streib
 * Altered POST cgi code to handle bug in Netscape
 * SO_SNDBUF changes by Larry

** Changes from 0.93.17.2 to 0.93.19 (all 0.93.18.x changes inclusive)
 * Update of some copyright statements for 99
 * Replacement of sprintf with strlen/memcpy or strcpy/strcat 
   wherever possible
 * Significant rearrangement in alias.c, minor functional differences
   (some CGI environment variables handled differently)
 * Removal of die function. Replace with log_err_mesg and exit.
 * initial IPv6 stubs and support
 * Move #include "config.h" to top of boa.h where it will do some good
 * Stubs and functions for strstr and strdup
 * Seperation of buffer code into it's own file
 * Significant changes to cgi.c et al (cgi_header.c, etc...)
 * Speed patches by removal of "extra" calls to time(): Use global variable!
 * pipelining changes... it works now.
 * require content-length from clients (ala rfc1945)
 * alter body_read and body_write to work more efficiently with known content-length
 * move read(2) part to *after* parsing...
 * added support for additional header message in send_redirect_temp
 * change use of NO_ZERO_FILL_LENGTH to offsetof() use
 * Remove SO_REUSEADDR setting on each client socket, Paul Saab <paul@mu.org>
 * Avoid SO_SNDBUF setting if possible
 * Large quantities of otherwise not-insignificant changes

** Changes from 0.93.17.2 to 0.93.17.3
 * Put on-the-fly directories back in, stripped down from the 0.92 version
 * Fixed DocumentRoot, ServerAdmin and ServerName null-value handling in
   CGI environment generation
 * Fixed argument order in Script* directives (bug introduced in 0.93.17.2)
 * Got rid of MAX_CGI_VARS because it was not being used consistently, or
   for that matter, at all, really.
 * Added some more FASCIST_LOGGING to cgi.c
 * Minor mmap patch by LRD for request.c

** Changes from 0.93.17.1 to 0.93.17.2
 * Added "Listen" directive for server bind address, as most recently
   suggested by David N. Welton <davidw@linuxcare.com>
 * Put virtualhost feature in, was experimental in 0.92q

** Changes from 0.93.16.2 to 0.93.17.1
 * New config file parser (supposed to be more maintainable) (LRD)
 * Support for "|command" and ":host:port" syntax for logfiles (untested) (LRD)

** Changes for the 0.93 version **
 * Huge quantities of changes
 * keepalive Bugfix in 0.93.16.2 by Jon Nelson
   report by Craig Silverstein of Google fame.
   (Craig Silverstein <csilvers@google.com>)
 * patch for config.h by Craig Silverstein <csilvers@google.com>
 * fixed "Parent Directory" problem in boa_indexer for title "/"
   (Debian bug #36165)
 * More Craig Silverstein <csilvers@google.com> modifications, namely:
   ErrorLog        (if omitted, print to stderr)
   DocumentRoot    (if omitted, can only server user-dir files)
   DirectoryIndex  (if omitted, always use DirectoryMaker)
   MimeTypes       (if omitted, don't load -- users can use AddType instead)


** Changes from v0.92o to v0.92p **

 * Documented misbehavior of CGI, SIGHUP, short aliases, stale dircache.
 * Documented how to patch signals.c for use on SunOS.
 * Closed file descriptor leak when redirecting a bare directory URL to
 one with an appended "/".
 * Closed potential file descriptor leak if errors encountered generating
 on-the-fly index.
 * Cleaned up include file handling to be simultaneously compatible with
 Linux, SunOS, HP-UX, and AIX.
 * Supress message body for codes 302, 400, 403, 404, 500, and 501 if
 incoming request is "HEAD".

** Changes from v0.91 to v0.92o **
 (0.92o released 27 December, 1996)

 * Maintenance handover from Paul Phillips to Larry Doolittle
 * Changed (char)NULL to '\0'
 * Cleaned up signal handler prototypes in signals.c
 * Modified handling of CGI environment variable PATH_TRANSLATED,
 should now work the same as NCSA.
 * More conservative buffer size in add_cgi_env()
 * Build argv list for a CGI script according to spec
 * Speedup process_header_line, eliminate potential memory leak

src/log.c
/*
 *  Boa, an http server
 *  Copyright (C) 1995 Paul Phillips <psp@well.com>
 *  Some changes Copyright (C) 1996 Larry Doolittle <ldoolitt@jlab.org>
 *  Some changes Copyright (C) 1999 Jon Nelson <jnelson@boa.org>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 1, or (at your option)
 *  any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

/* $Id: log.c,v 1.32 2000/05/26 02:42:41 jon Exp $*/

#include "boa.h"

FILE *access_log;

char *error_log_name;
char *access_log_name;
char *cgi_log_name;
int cgi_log_fd;

FILE *fopen_gen_fd(char *spec, const char *mode)
{
    int fd;
    if (!spec || *spec == '\0')
        return NULL;
    fd = open_gen_fd(spec);
    if (!fd)
        return NULL;
    return fdopen(fd, mode);
}

/*
 * Name: open_logs
 *
 * Description: Opens up the error log, ties it to stderr, and line
 * buffers it.
 */

void open_logs(void)
{
    int error_log;

    if (access_log_name) {
        /* Used the "a" flag with fopen, but fopen_gen_fd builds that in
         * implicitly when used as a file, and "a" is incompatible with
         * pipes and network sockets. */
        if (!(access_log = fopen_gen_fd(access_log_name, "w"))) {
            int errno_save = errno;
            fprintf(stderr, "Cannot open %s for logging: ",
                    access_log_name);
            errno = errno_save;
            perror("logfile open");
            exit(errno);
        }
        /* line buffer the access log */
#ifdef SETVBUF_REVERSED
        setvbuf(access_log, _IOLBF, (char *) NULL, 0);
#else
        setvbuf(access_log, (char *) NULL, _IOLBF, 0);
#endif
    } else
        access_log = NULL;

    if (cgi_log_name) {
        cgi_log_fd = open_gen_fd(cgi_log_name);
        if (cgi_log_fd == -1) {
            log_error_mesg(__FILE__, __LINE__, "open cgi_log");
            free(cgi_log_name);
            cgi_log_name = NULL;
            cgi_log_fd = 0;
        } else {
            if (fcntl(cgi_log_fd, F_SETFD, 1) == -1) {
                log_error_mesg(__FILE__, __LINE__,
                               "unable to set close-on-exec flag for cgi_log");
                close(cgi_log_fd);
                cgi_log_fd = 0;
                free(cgi_log_name);
                cgi_log_name = NULL;
            }
        }
    }

    /* if error_log_name is set, dup2 stderr to it */
    /* otherwise, leave stderr alone */
    if (error_log_name) {
        /* open the log file */
        if (!(error_log = open_gen_fd(error_log_name))) {
            log_error_mesg(__FILE__, __LINE__, "unable to open error log");
            exit(errno);
        }

        /* redirect stderr to error_log */
        if (dup2(error_log, STDERR_FILENO) == -1) {
            log_error_mesg(__FILE__, __LINE__,
                           "unable to dup2 the error log");
            exit(errno);
        }
        close(error_log);
    }

    /* set the close-on-exec to true */
    if (fcntl(STDERR_FILENO, F_SETFD, 1) == -1) {
        log_error_mesg(__FILE__, __LINE__,
                       "unable to fcntl the error log");
        exit(errno);
    }
}

/*
 * Name: close_access_log
 *
 * Description: closes access_log file
 */
void close_access_log(void)
{
    if (access_log)
        fclose(access_log);
}

/*
 * Name: log_access
 *
 * Description: Writes log data to access_log.
 */

void log_access(request * req)
{
    if (access_log) {
        if (virtualhost)
            fprintf(access_log, "%s ", req->local_ip_addr);
        fprintf(access_log, "%s - - %s\"%s\" %d %ld \"%s\" \"%s\"\n",
                req->remote_ip_addr,
                get_commonlog_time(),
                req->logline,
                req->response_status,
                req->filepos,
                (req->header_referer ? req->header_referer : "-"),
                (req->header_user_agent ? req->header_user_agent : "-"));

    }
}

/*
 * Name: log_error_doc
 *
 * Description: Logs the current time and transaction identification
 * to the stderr (the error log):
 * should always be followed by an fprintf to stderr
 *
 * This function used to be implemented with a big fprintf, but not
 * all fprintf's are reliable in the face of null string pointers
 * (SunOS, in particular).  As long as I had to add the checks for
 * null pointers, I changed from fprintf to fputs.
 *
 * Example output:
 [08/Nov/1997:01:05:03 -0600] request from 192.228.331.232 "GET /~joeblow/dir/ HTTP/1.0" ("/usr/user1/joeblow/public_html/dir/"): write: Broken pipe
 */

void log_error_doc(request * req)
{
    int errno_save = errno;

    fprintf(stderr, "%srequest from %s \"%s\" (\"%s\"): ",
            get_commonlog_time(),
            req->remote_ip_addr,
            (req->logline != NULL ?
             req->logline : "(null)"),
            (req->pathname != NULL ? req->pathname : "(null)"));

    errno = errno_save;
}

/*
 * Name: boa_perror
 *
 * Description: logs an error to user and error file both
 *
 */
void boa_perror(request * req, char *message)
{
    log_error_doc(req);
    perror(message);            /* don't need to save errno because log_error_doc does */
    send_r_error(req);
}

/*
 * Name: log_error_time
 *
 * Description: Logs the current time to the stderr (the error log):
 * should always be followed by an fprintf to stderr
 */

void log_error_time()
{
    int errno_save = errno;
    fputs(get_commonlog_time(), stderr);
    errno = errno_save;
}

/*
 * Name: log_error_mesg
 *
 * Description: performs a log_error_time, writes the file and lineno
 * to stderr (saving errno), and then a perror with message
 *
 */

void log_error_mesg(char *file, int line, char *mesg)
{
    int errno_save = errno;
    if (errno_save > sys_nerr - 1)
        fprintf(stderr, "%s%s:%d - %s:%s\n",
                get_commonlog_time(), file, line, "Unknown errno!!", mesg);
    else {
        fprintf(stderr, "%s%s:%d - ", get_commonlog_time(), file, line);
        errno = errno_save;
        perror(mesg);
    }
    errno = errno_save;
}


src/config.h
/* config.h.  Generated automatically by configure.  */
/* config.h.in.  Generated automatically from configure.in by autoheader.  */

/* Define to empty if the keyword does not work.  */
/* #undef const */

/* Define to `int' if <sys/types.h> doesn't define.  */
/* #undef gid_t */

/* Define if you have a working `mmap' system call.  */
/* #undef HAVE_MMAP */

/* Define if you have <sys/wait.h> that is POSIX.1 compatible.  */
#define HAVE_SYS_WAIT_H 1

/* Define to `int' if <sys/types.h> doesn't define.  */
/* #undef pid_t */

/* Define if the setvbuf function takes the buffering type as its second
   argument and the buffer pointer as the third, as on System V
   before release 3.  */
/* #undef SETVBUF_REVERSED */

/* Define if you have the ANSI C header files.  */
#define STDC_HEADERS 1

/* Define if you can safely include both <sys/time.h> and <time.h>.  */
#define TIME_WITH_SYS_TIME 1

/* Define if your <sys/time.h> declares struct tm.  */
/* #undef TM_IN_SYS_TIME */

/* Define to `int' if <sys/types.h> doesn't define.  */
/* #undef uid_t */

#define GUNZIP "/usr/bin/gunzip"

/* Define if you have the getcwd function.  */
#define HAVE_GETCWD 1

/* Define if you have the gethostname function.  */
#define HAVE_GETHOSTNAME 1

/* Define if you have the getpagesize function.  */
#define HAVE_GETPAGESIZE 1

/* Define if you have the select function.  */
#define HAVE_SELECT 1

/* Define if you have the socket function.  */
#define HAVE_SOCKET 1

/* Define if you have the strdup function.  */
#define HAVE_STRDUP 1

/* Define if you have the strstr function.  */
#define HAVE_STRSTR 1

/* Define if you have the <dirent.h> header file.  */
#define HAVE_DIRENT_H 1

/* Define if you have the <fcntl.h> header file.  */
#define HAVE_FCNTL_H 1

/* Define if you have the <limits.h> header file.  */
#define HAVE_LIMITS_H 1

/* Define if you have the <ndir.h> header file.  */
/* #undef HAVE_NDIR_H */

/* Define if you have the <sys/dir.h> header file.  */
/* #undef HAVE_SYS_DIR_H */

/* Define if you have the <sys/ndir.h> header file.  */
/* #undef HAVE_SYS_NDIR_H */

/* Define if you have the <sys/select.h> header file.  */
#define HAVE_SYS_SELECT_H 1

/* Define if you have the <sys/time.h> header file.  */
#define HAVE_SYS_TIME_H 1

/* Define if you have the <unistd.h> header file.  */
#define HAVE_UNISTD_H 1


src/defines.h
/*
 *  Boa, an http server
 *  Copyright (C) 1995 Paul Phillips <psp@well.com>
 *  Some changes Copyright (C) 1997 Jon Nelson <jnelson@boa.org>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 1, or (at your option)
 *  any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

/* $Id: defines.h,v 1.82 2000/05/01 20:33:25 jon Exp $*/

#ifndef _DEFINES_H
#define _DEFINES_H

/***** Change this, or use -c on the command line to specify it *****/

#ifndef SERVER_ROOT
#define SERVER_ROOT "/etc/boa"
#endif

/***** Various stuff that you may want to tweak, but probably shouldn't *****/

#define SOCKETBUF_SIZE				8192
#define MAX_HEADER_LENGTH			1024
#define CLIENT_STREAM_SIZE			SOCKETBUF_SIZE
#define BUFFER_SIZE				CLIENT_STREAM_SIZE

#define MIME_HASHTABLE_SIZE			47
#define ALIAS_HASHTABLE_SIZE                    17
#define PASSWD_HASHTABLE_SIZE		        47

#define REQUEST_TIMEOUT				60

#define CGI_MIME_TYPE    "application/x-httpd-cgi"
#define DEFAULT_PATH     "/bin:/usr/bin:/usr/local/bin"

/***** CHANGE ANYTHING BELOW THIS LINE AT YOUR OWN PERIL *****/
/***** You will probably introduce buffer overruns unless you know
       what you are doing *****/

#if defined HAVE_DIRENT_H
#include <dirent.h>             /* for MAXNAMLEN */
#elif defined HAVE_NDIR_H
#include <ndir.h>
#elif defined HAVE_SYS_DIR_H
#include <sys/dir.h>
#elif defined HAVE_SYS_NDIR_H
#include <sys/ndir.h>
#endif
#include <sys/socket.h>
#ifndef SO_MAXCONN
#define SO_MAXCONN 250
#endif

#define MAX_SITENAME_LENGTH			256
#define MAX_LOG_LENGTH				MAX_HEADER_LENGTH + 1024
#define MAX_FILE_LENGTH				NAME_MAX
#ifndef PATH_MAX
#define PATH_MAX 2048
#endif
#define MAX_PATH_LENGTH				PATH_MAX

#ifdef ACCEPT_ON
#define MAX_ACCEPT_LENGTH MAX_HEADER_LENGTH
#else
#define MAX_ACCEPT_LENGTH 0
#endif

#ifndef SERVER_VERSION
#define SERVER_VERSION 				"Boa/0.94.7"
#endif

#define CGI_VERSION				"CGI/1.1"
#define COMMON_CGI_COUNT 8
#define CGI_ENV_MAX     50
#define CGI_ARGC_MAX 128

/******************* RESPONSE CLASSES *****************/

#define R_INFORMATIONAL	1
#define R_SUCCESS	2
#define R_REDIRECTION	3
#define R_CLIENT_ERROR	4
#define R_SERVER_ERROR	5

/******************* RESPONSE CODES ******************/

#define R_REQUEST_OK	200
#define R_CREATED	201
#define R_ACCEPTED	202
#define R_PROVISIONAL	203       /* provisional information */
#define R_NO_CONTENT	204

#define R_MULTIPLE	300          /* multiple choices */
#define R_MOVED_PERM	301
#define R_MOVED_TEMP	302
#define R_NOT_MODIFIED	304

#define R_BAD_REQUEST	400
#define R_UNAUTHORIZED	401
#define R_PAYMENT	402           /* payment required */
#define R_FORBIDDEN	403
#define R_NOT_FOUND	404
#define R_METHOD_NA	405         /* method not allowed */
#define R_NONE_ACC	406          /* none acceptable */
#define R_PROXY		407            /* proxy authentication required */
#define R_REQUEST_TO	408        /* request timeout */
#define R_CONFLICT	409
#define R_GONE		410

#define R_ERROR		500            /* internal server error */
#define	R_NOT_IMP	501           /* not implemented */
#define	R_BAD_GATEWAY	502
#define R_SERVICE_UNAV	503      /* service unavailable */
#define	R_GATEWAY_TO	504        /* gateway timeout */
#define R_BAD_VERSION	505

/****************** METHODS *****************/

#define M_GET		1
#define M_HEAD		2
#define M_PUT		3
#define M_POST		4
#define M_DELETE	5
#define M_LINK		6
#define M_UNLINK	7

/******************* ERRORS *****************/

#define SERVER_ERROR		1
#define OUT_OF_MEMORY		2
#define NO_CREATE_SOCKET	3
#define NO_FCNTL		4
#define NO_SETSOCKOPT		5
#define NO_BIND			6
#define NO_LISTEN		7
#define NO_SETGID		8
#define NO_SETUID		9
#define NO_OPEN_LOG		10
#define SELECT			11
#define GETPWUID		12
#define INITGROUPS		13

#define SHUTDOWN		15            /* do not change */

/************** REQUEST STATUS (req->status) ***************/

#define READ_HEADER             0
#define ONE_CR                  1
#define ONE_LF                  2
#define TWO_CR                  3
#define BODY_READ               4
#define BODY_WRITE              5
#define WRITE                   6
#define PIPE_READ               7
#define PIPE_WRITE              8
#define DONE			9
#define DEAD                   10

#define CLIENT_WRITABLE(status) (status==WRITE || status==PIPE_WRITE)
#define CLIENT_READABLE(status) (status < BODY_WRITE)

/************** CGI TYPE (req->is_cgi) ******************/

#define CGI                     1
#define NPH                     2

/************* ALIAS TYPES (aliasp->type) ***************/

#define ALIAS			0
#define SCRIPTALIAS		1
#define REDIRECT		2

/*********** KEEPALIVE CONSTANTS (req->keepalive) *******/

#define KA_INACTIVE		0
#define KA_STOPPED     	1
#define KA_ACTIVE      	2

#define SQUASH_KA(req)	(req->keepalive=KA_STOPPED)

/********* CGI STATUS CONSTANTS (req->cgi_status) *******/
#define CGI_PARSE 1
#define CGI_BUFFER 2
#define CGI_DONE 3

/*********** MMAP_LIST CONSTANTS ************************/
#define MMAP_LIST_SIZE 256
#define MMAP_LIST_MASK 255
#define MMAP_LIST_USE_MAX 128
#define MMAP_LIST_NEXT(i) (((i)+1)&MMAP_LIST_MASK)
#define MMAP_LIST_HASH(dev,ino,size) ((ino)&MMAP_LIST_MASK)

#define MAX_FILE_MMAP 100 * 1024 /* 100K */

#ifndef NI_MAXHOST
#define NI_MAXHOST 20
#endif

#endif


src/cgi.c
/*
 *  Boa, an http server
 *  Copyright (C) 1995 Paul Phillips <psp@well.com>
 *  Some changes Copyright (C) 1996,97 Larry Doolittle <ldoolitt@jlab.org>
 *  Some changes Copyright (C) 1996 Charles F. Randall <crandall@goldsys.com>
 *  Some changes Copyright (C) 1997-99 Jon Nelson <jnelson@boa.org>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 1, or (at your option)
 *  any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

/* $Id: cgi.c,v 1.64 2000/05/26 02:42:58 jon Exp $ */

#include "boa.h"

int verbose_cgi_logs = 0;

static char **common_cgi_env;

/*
 * Name: create_common_env
 *
 * Description: Set up the environment variables that are common to
 * all CGI scripts
 */

void create_common_env()
{
    int index = 0;

    common_cgi_env =
        (char **) malloc(sizeof (char *) * COMMON_CGI_COUNT + 4);
    /* The +4 is to handle the extra 3 environment variables + the NULL in
       complete_env
     */

    /* NOTE NOTE NOTE:
       If you (the reader) someday modify this chunk of code to
       handle more "common" CGI environment variables, then bump the
       value in defines.h UP

       Also, in the case of document_root and server_admin, two variables
       that may or may not be defined depending on the way the server
       is configured, we check for null values and use an empty
       string to denote a NULL value to the environment, as per the
       specification. The quote for which follows:

       "In all cases, a missing environment variable is
       equivalent to a zero-length (NULL) value, and vice versa."
     */
    common_cgi_env[index++] = env_gen("PATH", DEFAULT_PATH);
    common_cgi_env[index++] = env_gen("SERVER_SOFTWARE", SERVER_VERSION);
    common_cgi_env[index++] = env_gen("SERVER_NAME", server_name);
    common_cgi_env[index++] = env_gen("GATEWAY_INTERFACE", CGI_VERSION);
    common_cgi_env[index++] =
        env_gen("SERVER_PORT", simple_itoa(server_port));

    /* NCSA and APACHE added -- not in CGI spec */
    common_cgi_env[index++] = env_gen("DOCUMENT_ROOT", document_root);

    /* NCSA added */
    common_cgi_env[index++] = env_gen("SERVER_ROOT", server_root);

    /* APACHE added */
    common_cgi_env[index++] = env_gen("SERVER_ADMIN", server_admin);
    common_cgi_env[index] = NULL;
}

/*
 * Name: create_env
 *
 * Description: Allocates memory for environment before execing a CGI
 * script.  I like spelling creat with an extra e, don't you?
 */

void create_env(request * req)
{
    int i;

    req->cgi_env = (char **) malloc(sizeof (char *) * CGI_ENV_MAX + 4);

    for (i = 0; common_cgi_env[i]; i++)
        req->cgi_env[i] = common_cgi_env[i];

    req->cgi_env_index = i;

    {
        char *w;

        switch (req->method) {
        case M_POST:
            w = "POST";
            break;
        case M_HEAD:
            w = "HEAD";
            break;
        case M_GET:
            w = "GET";
            break;
        default:
            w = "UNKNOWN";
            break;
        }
        req->cgi_env[req->cgi_env_index++] = env_gen("REQUEST_METHOD", w);
    }

    req->cgi_env[req->cgi_env_index++] =
        env_gen("SERVER_PROTOCOL", req->http_version);

    if (req->path_info) {
        req->cgi_env[req->cgi_env_index++] =
            env_gen("PATH_INFO", req->path_info);
        /* path_translated depends upon path_info */
        req->cgi_env[req->cgi_env_index++] =
            env_gen("PATH_TRANSLATED", req->path_translated);
    }
    req->cgi_env[req->cgi_env_index++] =
        env_gen("SCRIPT_NAME", req->script_name);

    if (req->query_string) {
        req->cgi_env[req->cgi_env_index++] =
            env_gen("QUERY_STRING", req->query_string);
    }
    req->cgi_env[req->cgi_env_index++] =
        env_gen("REMOTE_ADDR", req->remote_ip_addr);

    req->cgi_env[req->cgi_env_index++] =
        env_gen("REMOTE_PORT", simple_itoa(req->remote_port));
}

/*
 * Name: env_gen_extra
 *       (and via a not-so-tricky #define, env_gen)
 * This routine calls malloc: please free the memory when you are done
 */
char *env_gen_extra(const char *key, const char *value, int extra)
{
    char *result;
    int key_len, value_len;

    if (value == NULL)          /* ServerAdmin may not be defined, eg */
        value = "";
    key_len = strlen(key);
    value_len = strlen(value);
    /* leave room for '=' sign and null terminator */
    result = malloc(extra + key_len + value_len + 2);
    if (result) {
        memcpy(result + extra, key, key_len);
        *(result + extra + key_len) = '=';
        memcpy(result + extra + key_len + 1, value, value_len);
        *(result + extra + key_len + value_len + 1) = '\0';
    }
    return result;
}

/*
 * Name: add_cgi_env
 *
 * Description: adds a variable to CGI's environment
 * Used for HTTP_ headers
 */

void add_cgi_env(request * req, char *key, char *value)
{
    char *p;

    if (req->cgi_env_index < CGI_ENV_MAX) {
        p = env_gen_extra(key, value, 5);
        memcpy(p, "HTTP_", 5);
        req->cgi_env[req->cgi_env_index++] = p;
    } else {
        log_error_time();
        fprintf(stderr, "Unable to generate additional CGI Environment"
                "variable -- not enough space!\n");
    }
}

/*
 * Name: complete_env
 *
 * Description: adds the known client header env variables
 * and terminates the environment array
 */

void complete_env(request * req)
{
    if (req->method == M_POST) {
        if (req->content_type)
            req->cgi_env[req->cgi_env_index++] =
                env_gen("CONTENT_TYPE", req->content_type);
        else
            req->cgi_env[req->cgi_env_index++] =
                env_gen("CONTENT_TYPE", default_type);

        if (req->content_length) {
            req->cgi_env[req->cgi_env_index++] =
                env_gen("CONTENT_LENGTH", req->content_length);
        }
    }
#ifdef ACCEPT_ON
    if (req->accept[0]) {
        req->cgi_env[req->cgi_env_index++] =
            env_gen("HTTP_ACCEPT", req->accept);
    }
#endif
    req->cgi_env[req->cgi_env_index] = NULL; /* terminate */
}

/*
 * Name: make_args_cgi
 *
 * Build argv list for a CGI script according to spec
 *
 */

void create_argv(request * req, char **aargv)
{
    char *p, *q, *r;
    int aargc;

    q = req->query_string;
    aargv[0] = req->pathname;

    if (q && !strchr(q, '=')) {
        /* fprintf(stderr,"Parsing string %s\n",q); */
        q = strdup(q);
        for (aargc = 1; q && (aargc < CGI_ARGC_MAX);) {
            r = q;
            if ((p = strchr(q, '+'))) {
                *p = '\0';
                q = p + 1;
            } else {
                q = NULL;
            }
            if (unescape_uri(r)) {
                /* printf("parameter %d: %s\n",aargc,r); */
                aargv[aargc++] = r;
            }
        }
        aargv[aargc] = NULL;
    } else {
        aargv[1] = NULL;
    }
}

/*
 * Name: init_cgi
 *
 * Description: Called for GET/POST requests that refer to ScriptAlias
 * directories or application/x-httpd-cgi files.  Ties stdout to socket,
 * stdin to data if POST, and execs CGI.
 * stderr remains tied to our log file; is this good?
 *
 * Returns:
 * 0 - error or NPH, either way the socket is closed
 * 1 - success
 */

int init_cgi(request * req)
{
    int child_pid;
    int pipes[2];

    SQUASH_KA(req);

    if (req->is_cgi)
        complete_env(req);
#ifdef FASCIST_LOGGING
    {
        int i;
        for (i = 0; i < req->cgi_env_index; ++i)
            fprintf(stderr, "%s - environment variable for cgi: \"%s\"\n",
                    __FILE__, req->cgi_env[i]);
    }
#endif

    if (req->is_cgi == CGI)
        pipe(pipes);

    if ((child_pid = fork()) == -1) { /* fork unsuccessful */
        log_error_time();
        perror("fork");

        if (req->is_cgi == CGI) {
            close(pipes[0]);
            close(pipes[1]);
        }
        send_r_error(req);
        /* FIXME: There is aproblem here. send_r_error would work
           for NPH and CGI, but not for GUNZIP.  Fix that. */
        /* i'd like to send_r_error, but.... */
        return 0;
    }
    /* if here, fork was successful */

    if (!child_pid) {           /* 0 == child */
        if (req->is_cgi == CGI) {
            /* tie cgi's STDOUT to it's write end of pipe */
            if (dup2(pipes[1], STDOUT_FILENO) == -1) {
                log_error_time();
                perror("dup2 - pipes");
                close(pipes[0]);
                close(pipes[1]);
                exit(1);
            }
            if (fcntl(pipes[1], F_SETFL, 0) == -1) {
                log_error_time();
                perror("cgi-fcntl");
                exit(1);
            }
        } else {
            /* tie stdout to socket */
            if (dup2(req->fd, STDOUT_FILENO) == -1) {
                log_error_time();
                perror("dup2 - fd");
                exit(1);
            }
            /* Switch socket flags back to blocking */
            if (fcntl(req->fd, F_SETFL, 0) == -1) {
                log_error_time();
                perror("cgi-fcntl");
                exit(1);
            }
        }
        /* tie post_data_fd to POST stdin */
        if (req->method == M_POST) { /* tie stdin to file */
            lseek(req->post_data_fd, SEEK_SET, 0);
            dup2(req->post_data_fd, STDIN_FILENO);
            close(req->post_data_fd);
        }
        /* Close access log, so CGI program can't scribble
         * where it shouldn't
         */
        close_access_log();

        /*
         * tie STDERR to cgi_log_fd
         * cgi_log_fd will automatically close, close-on-exec rocks!
         * if we don't tied STDERR (current log_error) to cgi_log_fd,
         *  then we ought to close it.
         */
        if (!cgi_log_fd)
            close(STDERR_FILENO);
        else
            dup2(cgi_log_fd, STDERR_FILENO);

        if (req->is_cgi) {
            char *aargv[CGI_ARGC_MAX + 1];
            create_argv(req, aargv);
            execve(req->pathname, aargv, req->cgi_env);
        } else {
            if (req->pathname[strlen(req->pathname) - 1] == '/')
                execl(dirmaker, dirmaker, req->pathname, req->request_uri,
                      NULL);
            else
                execl(GUNZIP, GUNZIP, "--stdout", "--decompress",
                      req->pathname, NULL);
        }
        /* execve failed */
        log_error_mesg(__FILE__, __LINE__, req->pathname);
        exit(1);
    }

    /* if here, fork was successful */
    if (verbose_cgi_logs) {
        log_error_time();
        fprintf(stderr, "Forked child \"%s\" pid %d\n",
                req->pathname, child_pid);
    }

    if (req->method == M_POST)
        close(req->post_data_fd); /* child closed it too */

    /* NPH, GUNZIP, etc... all go straight to the fd */
    if (req->is_cgi != CGI)
        return 0;

    close(pipes[1]);
    req->data_fd = pipes[0];

    req->status = PIPE_READ;
    if (req->is_cgi == CGI) {
        req->cgi_status = CGI_PARSE; /* got to parse cgi header */
        /* for cgi_header... I get half the buffer! */
        req->header_line = req->header_end =
            (req->buffer + BUFFER_SIZE / 2);
    } else {
        req->cgi_status = CGI_BUFFER;
        /* I get all the buffer! */
        req->header_line = req->header_end = req->buffer;
    }

    /* reset req->filepos for logging (it's used in pipe.c) */
    /* still don't know why req->filesize might be reset though */
    req->filepos = 0;

    /* set the read end of the socket to non-blocking */
    if (fcntl(pipes[0], F_SETFL, O_NONBLOCK) == -1) {
        log_error_time();
        perror("cgi-fcntl");
        return 0;
    }

    return 1;
}
